pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSink.java

```diff
@@ -0,0 +1,156 @@
+ * a NiFiDataPacketBuilder which can create instances of NiFiDataPacket from the incoming data.
+ */
+@Connector(
+    name = "nifi",
+    type = IOType.SINK,
+    help = "The NiFiSink is used for moving messages from Pulsar to Apache NiFi using the NiFi Site-to-Site client.",
+    configClass = NiFiConfig.class
+)
+@Slf4j
+public class NiFiSink implements Sink<NiFiDataPacket> {
+
+    private NiFiConfig niFiConfig;
+    private SiteToSiteClientConfig clientConfig;
+    private SiteToSiteClient client;
+
+    private int requestBatchCount;
+    private long waitTimeMs;
+    private List<Record<NiFiDataPacket>>  currentList;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        niFiConfig = NiFiConfig.load(config);
+        Preconditions.checkNotNull(niFiConfig.getUrl(), "url property not set.");
+        Preconditions.checkNotNull(niFiConfig.getPortName(), "portName property not set.");
+        Preconditions.checkArgument(niFiConfig.getRequestBatchCount() > 0,
+                "requestBatchCount must be a positive integer.");
+
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+        client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+
+        requestBatchCount = niFiConfig.getRequestBatchCount();
+        waitTimeMs = niFiConfig.getWaitTimeMs();
+        currentList= Lists.newArrayList();
+
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), waitTimeMs, waitTimeMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (null != client) {
+            client.close();
+        }
+
+        if (null != flushExecutor) {
+            flushExecutor.shutdown();
+        }
+
+    }
+
+    @Override
+    public void write(Record<NiFiDataPacket> record) {
+        int number;
+        synchronized (this) {
+            if (null != record) {
+                currentList.add(record);
+            }
+            number = currentList.size();
+        }
+
+        if (number == requestBatchCount) {
+            flushExecutor.submit(() -> flush());
+        }
+    }
+
+    private void flush() {
+        List<Record<NiFiDataPacket>>  toFlushList;
+        synchronized (this) {
+            if (currentList.isEmpty()) {
+                return;
+            }
+            toFlushList = currentList;
+            currentList = Lists.newArrayList();
+        }
+
+        Transaction transaction = null;
+        try {
+            transaction = client.createTransaction(TransferDirection.SEND);
+        } catch (IOException ioe) {
+            log.warn("Created NiFi transaction Failed", ioe);
```

Comment:
if we failed to create transaction, we have to error out the `toFlushList`.

Reply:
> if we failed to create transaction, we have to error out the `toFlushList`.

if (null != transaction) {
} else {
            log.error("Failed to send records", toFlushList);
 }